File: common/src/main/java/com/jon/common/utils/Protocol.java
package com.jon.common.utils;

import android.content.SharedPreferences;

public enum Protocol {
    SSL("SSL"),
    TCP("TCP"),
    UDP("UDP");

    private final String protocol;

    Protocol(String protocol) {
        this.protocol = protocol;
    }

    public String get() {
        return protocol;
    }

    public static Protocol fromPrefs(SharedPreferences prefs) {
        final String protocolString = PrefUtils.getString(prefs, Key.TRANSMISSION_PROTOCOL);
        return fromString(protocolString);
    }

    public static Protocol fromString(String protocolString) {
        for (Protocol protocol : values()) {
            if (protocol.get().equals(protocolString)) {
                return protocol;
            }
        }
        throw new IllegalArgumentException("Unknown protocol: " + protocolString);
    }
}

File: common/src/main/java/com/jon/common/utils/Protocol.kt
package com.jon.common.utils

import android.annotation.SuppressLint
import android.content.SharedPreferences
import com.jon.common.prefs.CommonPrefs
import com.jon.common.prefs.PrefPair
import com.jon.common.prefs.getStringFromPair

enum class Protocol(
        private val protocol: String,
        val presetPref: PrefPair<String>
) {
    SSL(
            protocol = "SSL",
            presetPref = CommonPrefs.SSL_PRESETS
    ),
    TCP(
            protocol = "TCP",
            presetPref = CommonPrefs.TCP_PRESETS
    ),
    UDP(
            protocol = "UDP",
            presetPref = CommonPrefs.UDP_PRESETS
    );

    override fun toString(): String {
        return protocol
    }

    companion object {
        fun fromPrefs(prefs: SharedPreferences): Protocol {
            val protocolString = prefs.getStringFromPair(CommonPrefs.TRANSMISSION_PROTOCOL)
            return fromString(protocolString)
        }

        @SuppressLint("DefaultLocale")
        fun fromString(protocolString: String): Protocol {
            return values().firstOrNull { it.toString().toLowerCase() == protocolString.toLowerCase() }
                    ?: throw IllegalArgumentException("Unknown protocol: $protocolString")
        }
    }
}
